File: native/ios/Sources/NativeJourney.swift
import Foundation

struct JourneyChapter: Decodable, Identifiable {
    let id: String
    let label: String
    let title: String
    let emphasis: String
    let body: String
}

struct NativeJourney {
    let editorial: NativeEditorial
    let brand: String
    let notice: String
    let hero: JourneyChapter
    let login: String
    let home: String
    let illustration: String
    let imageDescription: String
    let poster: String
    let homePoster: String
    let chapters: [JourneyChapter]

    init(data: Data) throws {
        editorial = try NativeEditorial(data: data)
        struct Hero: Decodable { let eyebrow: String; let title: String; let emphasis: String; let body: String; let illustration: String }
        struct Navigation: Decodable { let login: String; let home: String }
        struct Media: Decodable { let poster: String; let homePoster: String; let alt: String }
        struct Document: Decodable {
            let brand: String; let notice: String; let hero: Hero
            let navigation: Navigation; let media: Media; let stages: [JourneyChapter]
        }
        let document = try JSONDecoder().decode(Document.self, from: data)
        let values = [document.brand, document.notice, document.hero.eyebrow, document.hero.title,
                      document.hero.emphasis, document.hero.body, document.hero.illustration,
                      document.navigation.login, document.navigation.home, document.media.alt] +
            document.stages.flatMap { [$0.id, $0.label, $0.title, $0.emphasis, $0.body] }
        guard values.allSatisfy({ !$0.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty }),
              !document.stages.isEmpty, Set(document.stages.map(\.id)).count == document.stages.count else {
            throw MobileFailure.configuration
        }
        func asset(_ path: String) throws -> String {
            guard path.range(of: "^/media/journey/[A-Za-z0-9_-]+\\.webp$", options: .regularExpression) != nil else {
                throw MobileFailure.configuration
            }
            return (path as NSString).lastPathComponent
        }
        brand = document.brand; notice = document.notice
        hero = JourneyChapter(id: "hero", label: document.hero.eyebrow, title: document.hero.title,
                              emphasis: document.hero.emphasis, body: document.hero.body)
        login = document.navigation.login; home = document.navigation.home
        illustration = document.hero.illustration; imageDescription = document.media.alt
        poster = try asset(document.media.poster); homePoster = try asset(document.media.homePoster)
        chapters = document.stages
    }
}

File: native/android/app/src/main/java/com/mediconnect/nativeapp/NativeJourney.kt
package com.mediconnect.nativeapp

import org.json.JSONObject

data class JourneyChapter(val id: String, val label: String, val title: String, val emphasis: String, val body: String)

class NativeJourney(document: JSONObject) {
    val editorial = NativeEditorial(document)
    val demo = NativeDemo(document, editorial)
    private fun text(objectValue: JSONObject, key: String): String = (objectValue.get(key) as? String)
        ?.also { require(it.isNotBlank()) } ?: throw IllegalArgumentException("Invalid journey content")
    val brand = text(document, "brand")
    val notice = text(document, "notice")
    val hero = document.getJSONObject("hero").let {
        JourneyChapter("hero", text(it, "eyebrow"), text(it, "title"), text(it, "emphasis"), text(it, "body"))
    }
    val login = text(document.getJSONObject("navigation"), "login")
    val home = text(document.getJSONObject("navigation"), "home")
    val illustration = text(document.getJSONObject("hero"), "illustration")
    val imageDescription = text(document.getJSONObject("media"), "alt")
    val poster = asset(text(document.getJSONObject("media"), "poster"))
    val homePoster = asset(text(document.getJSONObject("media"), "homePoster"))
    val chapters = document.getJSONArray("stages").let { stages ->
        List(stages.length()) { index -> stages.getJSONObject(index).let {
            JourneyChapter(text(it, "id"), text(it, "label"), text(it, "title"), text(it, "emphasis"), text(it, "body"))
        } }.also { require(it.isNotEmpty() && it.map { chapter -> chapter.id }.distinct().size == it.size) }
    }
    private fun asset(path: String): String {
        require(Regex("^/media/journey/[A-Za-z0-9_-]+\\.webp$").matches(path))
        return path.substringAfterLast('/')
    }
}
